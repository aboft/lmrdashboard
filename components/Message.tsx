import { IUser } from "./User";

export default function Message(props: IMessage) {
    return <div>[{props.time}] {props.user && props.user.nick}: {props.text}</div>;
}

export interface IMessage {
    text: string;
    user?: IUser;
    time: string;
}
