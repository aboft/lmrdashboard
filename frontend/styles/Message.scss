@use 'Srcery';

a {
  color: Srcery.$blue;
}
