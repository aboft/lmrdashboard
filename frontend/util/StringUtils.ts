export const StringUtils = {

  hasTrailingChar: (text: string): boolean => {
    return [',', '.', '!', '?'].includes(text[text.length - 1])
  },

  splitToTextWithoutTrailingAndTrailing: (text: string): Array<string>=> {
    const textWithoutEndChar = text.slice(0, text.length - 1);
    const endChar = text[text.length - 1];
    return [textWithoutEndChar, endChar];
  }
}
