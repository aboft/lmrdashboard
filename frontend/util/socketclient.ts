import openSocket from 'socket.io-client';

const socket = openSocket('http://localhost:8000'); function subscribeToTimer(cb: (timestamp: Date) => void) {
    socket.on('timer', (timestamp: Date) => cb(timestamp));
    socket.emit('subscribeToTimer', 5000);
}

export { subscribeToTimer };