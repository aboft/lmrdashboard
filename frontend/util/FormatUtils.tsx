import innerText from "react-innertext";
import { StringUtils } from "./StringUtils";

export const FormatUtils = {
  /**
   * Formats everything between (and including) two asterisks bold
   * Assumes only one pair of asterisks in the text
   */
  formatBoldViaAsterisks: (text: string): JSX.Element => {
    const previousText = text.substring(0, text.indexOf('*'));
    const toMakeBold = text.substring(text.indexOf('*'), text.lastIndexOf('*') + 1);
    const restOfTheText = text.substring(text.lastIndexOf('*') + 1);
    return <> {previousText}<b>{toMakeBold}</b>{restOfTheText}</>;
  },

  /**
   * Formats everything after the bold escape character bold
   */
  formatBoldViaEscapeCharacter: (preparedMessage: JSX.Element[]): JSX.Element[] => {
    const formattedMessage: JSX.Element[] = [];
    let escapeCharUsed = false;

    preparedMessage.forEach((messagePart: JSX.Element) => {
      const messagePartInnerText: string = innerText(messagePart);

      // Always apply the same formatting once an escape character has been used
      // (No un-escaping or other escape chars supported)
      if (escapeCharUsed) {
        formattedMessage.push(<b>{messagePart}</b>);
      } else if (messagePartInnerText.includes('\u0002')) {
        formattedMessage.push(<b>{messagePart}</b>);
        escapeCharUsed = true;
      } else {
        formattedMessage.push(<>{messagePart}</>);
      }
    });

    return formattedMessage;

  },

  /**
   * Formats link depending on whether the href or clickable text contains trailing characters.
   * Also splits up the href and clickable into parts with and without the trailing characters.
   */
  formatLink: (href: string, clickable: string): JSX.Element => {
    if (StringUtils.hasTrailingChar(clickable) || StringUtils.hasTrailingChar(href)) {
      const [clickableWithoutEndChar, endChar] = StringUtils.splitToTextWithoutTrailingAndTrailing(clickable);
      const hrefWithoutEndChar = StringUtils.splitToTextWithoutTrailingAndTrailing(href)[0];
      return <> <a href={hrefWithoutEndChar}>{clickableWithoutEndChar}</a>{endChar}</>;
    }
    return <> <a href={href}>{clickable}</a></>;
  }
}
