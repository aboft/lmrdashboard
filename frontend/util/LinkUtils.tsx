import { StringUtils } from "./StringUtils";

export const LinkUtils = {
  /**
   * Formats link depending on whether the href or clickable text contains trailing characters.
   * Also splits up the href and clickable into parts with and without the trailing characters.
   */
  formatLink: (href: string, clickable: string): JSX.Element => {
    if (StringUtils.hasTrailingChar(clickable) || StringUtils.hasTrailingChar(href)) {
      const [clickableWithoutEndChar, endChar] = StringUtils.splitToTextWithoutTrailingAndTrailing(clickable);
      const hrefWithoutEndChar = StringUtils.splitToTextWithoutTrailingAndTrailing(href)[0];
      return <> <a href={hrefWithoutEndChar}>{clickableWithoutEndChar}</a>{endChar}</>;
    }
    return <> <a href={href}>{clickable}</a></>;
  }
}
