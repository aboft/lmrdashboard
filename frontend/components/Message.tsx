import { getNickCSSClass } from '../data/UserHash';
import { LinkUtils } from '../util/LinkUtils';

export default function Message(props: IMessage): JSX.Element {

  /** 
   * If a message contains URLs or (/)r/<subreddit> those parts will be linked.
   */
  function linkifyMessageContent(message: string): JSX.Element[] {
    return message.split(' ')
      .map(messagePart => {
        if (messagePart.startsWith('r/')) {
          return LinkUtils.formatLink('https://reddit.com/' + messagePart, messagePart);
        } else if (messagePart.startsWith('/r/')) {
          return LinkUtils.formatLink('https://reddit.com' + messagePart, messagePart);
        } else if (messagePart.startsWith('https://')) {
          return LinkUtils.formatLink(messagePart, messagePart);
        } else {
          return <> {messagePart}</>
        }
      });
  }

  function linkifyMessageContentIfMessageContainsUrlOrSubreddit(message: string): string | JSX.Element[] {
    if (/^.*((https:\/\/)|(\/?r\/)).*$/.test(message)) {
      return linkifyMessageContent(message);
    }
    return message;
  }

  /**
   * Returns a message formatted in the way irssi formats it for /me.
   */
  function getFormattedActionMessage(): JSX.Element {
    return (<>
      [{props.dateCreated}] * <span className={getNickCSSClass(props.nick)}>{props.nick}</span>{props.message.replace('ACTION', '')}
    </>)
  }

  /**
   * Returns a message formatted in a normal <nick>: <message> format.
   */
  function getFormattedNormalMessage(): JSX.Element {
    return (<>
      [{props.dateCreated}] <span className={getNickCSSClass(props.nick)}>{props.nick}</span>: {
        linkifyMessageContentIfMessageContainsUrlOrSubreddit(props.message)
      }
    </>)
  }

  return (
    <div>
      {props.message.slice(1).startsWith('ACTION') ?
        getFormattedActionMessage() :
        getFormattedNormalMessage()}
    </div>
  );
}

export interface IMessage {
  nick: string;
  message: string;
  dateCreated: string;
}
