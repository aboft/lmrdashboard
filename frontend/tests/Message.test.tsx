import React from 'react';
import renderer from 'react-test-renderer';
import { expect } from '@jest/globals';

import Message from '../components/Message';
import { FormatUtils } from '../util/FormatUtils';

afterEach(() => {
  jest.clearAllMocks();
});

describe('no formatting', () => {
  test('message without links does not have special formatting', () => {
    const linkSpy = jest.spyOn(FormatUtils, 'formatLink');
    const boldSpy = jest.spyOn(FormatUtils, 'formatBoldViaAsterisks');

    const component = renderer.create(
      <Message nick={'Linus'} message={'This is a message without links.'} dateCreated={'2021-01-01'} />
    );

    const tree = component.toJSON();
    expect(linkSpy).toHaveBeenCalledTimes(0);
    expect(boldSpy).toHaveBeenCalledTimes(0);
    expect(tree).toMatchSnapshot();
  });
});

describe('link formatting', () => {
  test('message that contains an "r" but no links does not trigger link rendering', () => {
    const spy = jest.spyOn(FormatUtils, 'formatLink');

    const component = renderer.create(
      <Message
        nick={'Linus'}
        message={'This is a message without links, but it has an r.'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(spy).toHaveBeenCalledTimes(0);
    expect(tree).toMatchSnapshot();
  });

  test('message with URLs has links rendered', () => {
    const spy = jest.spyOn(FormatUtils, 'formatLink');

    const component = renderer.create(
      <Message
        nick={'RMS'}
        message={'This is a message with: http://reddit.com, also https://reddit.com/r/linuxmasterrace'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(spy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  test('message with subreddits has them rendered as links', () => {
    const spy = jest.spyOn(FormatUtils, 'formatLink');

    const component = renderer.create(
      <Message
        nick={'Gnu'}
        message={'This is a message with: r/linuxmasterrace, also /r/linux'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(spy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  test('message with mixed link types has them rendered', () => {
    const spy = jest.spyOn(FormatUtils, 'formatLink');
    const component = renderer.create(
      <Message
        nick={'Gnu'}
        message={'This is a message with: r/linuxmasterrace, which links to https://reddit.com/r/linuxmasterrace.'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(spy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });
});

describe('bold formatting', () => {
  test('message with asterisks has bold formatting', () => {
    const spy = jest.spyOn(FormatUtils, 'formatBoldViaAsterisks');
    const component = renderer.create(
      <Message
        nick={'BoldMoveCotton'}
        message={'This is a *message* with asterisk *bolded* words, *right*?'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(spy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  test('message with asterisks and links has both formatting', () => {
    const linkSpy = jest.spyOn(FormatUtils, 'formatLink');
    const boldSpy = jest.spyOn(FormatUtils, 'formatBoldViaAsterisks');

    const component = renderer.create(
      <Message
        nick={'BoldMoveCotton'}
        message={'This is a message with asterisk *bolded* words and links: https://reddit.com'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(linkSpy).toHaveBeenCalled();
    expect(boldSpy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  test('message with bold escape code has bold formatting', () => {
    const asteriskSpy = jest.spyOn(FormatUtils, 'formatBoldViaAsterisks');
    const escapeSpy = jest.spyOn(FormatUtils, 'formatBoldViaEscapeCharacter');

    const component = renderer.create(
      <Message
        nick={'BoldMoveCotton'}
        message={'This is a message with \u0002escape codes'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(asteriskSpy).toHaveBeenCalledTimes(0);
    expect(escapeSpy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  test('message with bold escape code(s) can be unescaped', () => {
    const escapeSpy = jest.spyOn(FormatUtils, 'formatBoldViaEscapeCharacter');

    const component = renderer.create(
      <Message
        nick={'BoldMoveCotton'}
        message={'\u0002bold\u0002 not bold \u0002bold again'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(escapeSpy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  test('message with bold escape code(s) escaping multiple words can be unescaped', () => {
    const escapeSpy = jest.spyOn(FormatUtils, 'formatBoldViaEscapeCharacter');

    const component = renderer.create(
      <Message
        nick={'BoldMoveCotton'}
        message={'notbold \u0002both bold\u0002 not bold \u0002bold again'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(escapeSpy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  test('message with asterisks and bold escape code has bold formatting', () => {
    const asteriskSpy = jest.spyOn(FormatUtils, 'formatBoldViaAsterisks');
    const escapeSpy = jest.spyOn(FormatUtils, 'formatBoldViaEscapeCharacter');

    const component = renderer.create(
      <Message
        nick={'BoldMoveCotton'}
        message={'This is a message with *asterisks* and \u0002escape codes'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(asteriskSpy).toHaveBeenCalled();
    expect(escapeSpy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });

  test('message with asterisks, bold escape code and links has correct formatting', () => {
    const asteriskSpy = jest.spyOn(FormatUtils, 'formatBoldViaAsterisks');
    const escapeSpy = jest.spyOn(FormatUtils, 'formatBoldViaEscapeCharacter');
    const linkSpy = jest.spyOn(FormatUtils, 'formatLink');

    const component = renderer.create(
      <Message
        nick={'BoldMoveCotton'}
        message={'This is a message with *asterisks*, \u0002escape codes and http://reddit.com'}
        dateCreated={'2021-01-01'}
      />
    );

    const tree = component.toJSON();
    expect(asteriskSpy).toHaveBeenCalled();
    expect(escapeSpy).toHaveBeenCalled();
    expect(linkSpy).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
  });
});
