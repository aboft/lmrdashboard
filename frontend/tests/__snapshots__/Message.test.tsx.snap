// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bold formatting message with asterisks and bold escape code has bold formatting 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-1"
  >
    BoldMoveCotton
  </span>
  : 
   
  This
   
  is
   
  a
   
  message
   
  with
   
  
  <b>
    *asterisks*
  </b>
  
   
  and
  <b>
     
    escape
  </b>
  <b>
     
    codes
  </b>
</div>
`;

exports[`bold formatting message with asterisks and links has both formatting 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-1"
  >
    BoldMoveCotton
  </span>
  : 
   
  This
   
  is
   
  a
   
  message
   
  with
   
  asterisk
   
  
  <b>
    *bolded*
  </b>
  
   
  words
   
  and
   
  links:
   
  <a
    href="https://reddit.com"
  >
    https://reddit.com
  </a>
</div>
`;

exports[`bold formatting message with asterisks has bold formatting 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-1"
  >
    BoldMoveCotton
  </span>
  : 
   
  This
   
  is
   
  a
   
  
  <b>
    *message*
  </b>
  
   
  with
   
  asterisk
   
  
  <b>
    *bolded*
  </b>
  
   
  words,
   
  
  <b>
    *right*
  </b>
  ?
</div>
`;

exports[`bold formatting message with asterisks, bold escape code and links has correct formatting 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-1"
  >
    BoldMoveCotton
  </span>
  : 
   
  This
   
  is
   
  a
   
  message
   
  with
   
  
  <b>
    *asterisks*
  </b>
  ,
  <b>
     
    escape
  </b>
  <b>
     
    codes
  </b>
  <b>
     
    and
  </b>
  <b>
     
    <a
      href="http://reddit.com"
    >
      http://reddit.com
    </a>
  </b>
</div>
`;

exports[`bold formatting message with bold escape code has bold formatting 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-1"
  >
    BoldMoveCotton
  </span>
  : 
   
  This
   
  is
   
  a
   
  message
   
  with
  <b>
     
    escape
  </b>
  <b>
     
    codes
  </b>
</div>
`;

exports[`link formatting message that contains an "r" but no links does not trigger link rendering 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-15"
  >
    Linus
  </span>
  : 
  This is a message without links, but it has an r.
</div>
`;

exports[`link formatting message with URLs has links rendered 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-8"
  >
    RMS
  </span>
  : 
   
  This
   
  is
   
  a
   
  message
   
  with:
   
  <a
    href="http://reddit.com"
  >
    http://reddit.com
  </a>
  ,
   
  also
   
  <a
    href="https://reddit.com/r/linuxmasterrace"
  >
    https://reddit.com/r/linuxmasterrace
  </a>
</div>
`;

exports[`link formatting message with mixed link types has them rendered 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-14"
  >
    Gnu
  </span>
  : 
   
  This
   
  is
   
  a
   
  message
   
  with:
   
  <a
    href="https://reddit.com/r/linuxmasterrace"
  >
    r/linuxmasterrace
  </a>
  ,
   
  which
   
  links
   
  to
   
  <a
    href="https://reddit.com/r/linuxmasterrace"
  >
    https://reddit.com/r/linuxmasterrace
  </a>
  .
</div>
`;

exports[`link formatting message with subreddits has them rendered as links 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-14"
  >
    Gnu
  </span>
  : 
   
  This
   
  is
   
  a
   
  message
   
  with:
   
  <a
    href="https://reddit.com/r/linuxmasterrace"
  >
    r/linuxmasterrace
  </a>
  ,
   
  also
   
  <a
    href="https://reddit.com/r/linux"
  >
    /r/linux
  </a>
</div>
`;

exports[`no formatting message without links does not have special formatting 1`] = `
<div>
  [
  2021-01-01
  ] 
  <span
    className="nick-15"
  >
    Linus
  </span>
  : 
  This is a message without links.
</div>
`;
